merkle/compact/range.go

```diff
@@ -0,0 +1,260 @@
+//
+//            [2 0]
+//           /     \
+//        [1 0]   [1 1]
+//       /   \     /   \
+//   [0 0] [0 1] [0 2] [0 3]
+type VisitFn func(level uint, index uint64, hash []byte)
+
+// RangeFactory allows creating compact ranges with the specified hash
+// function, which must not be nil, and must not be changed.
+type RangeFactory struct {
+	Hash HashFn
+}
+
+// NewRange creates a Range for [begin, end) with the given set of hashes. The
+// hashes correspond to the roots of the minimal set of perfect sub-trees
+// covering the [begin, end) leaves range, ordered left to right.
+func (f *RangeFactory) NewRange(begin, end uint64, hashes [][]byte) (*Range, error) {
+	if end < begin {
+		return nil, fmt.Errorf("invalid range: end=%d, want >= %d", end, begin)
+	}
+	left, right := decompose(begin, end)
+	ones := bits.OnesCount64(left) + bits.OnesCount64(right)
+	if ln := len(hashes); ln != ones {
+		return nil, fmt.Errorf("invalid hashes: got %d values, want %d", ln, ones)
+	}
+	return &Range{f: f, begin: begin, end: end, hashes: hashes}, nil
+}
+
+// NewEmptyRange returns a new Range for an empty [begin, begin) range. The
+// value of begin defines where the range will start growing from when entries
+// are appended to it.
+func (f *RangeFactory) NewEmptyRange(begin uint64) *Range {
+	return &Range{f: f, begin: begin, end: begin}
+}
+
+// Range represents a compact Merkle tree range for leaf indices [begin, end).
+//
+// It contains the minimal set of perfect subtrees whose leaves comprise this
+// range. The structure is efficiently mergeable with with other compact ranges
+// that share one of the endpoints with it.
+//
+// TODO(pavelkalinnikov): Add document with more details on how it works, and
+// what it can be used for.
+type Range struct {
+	f      *RangeFactory
+	begin  uint64
+	end    uint64
+	hashes [][]byte
+}
+
+// Begin returns the first index covered by the range (inclusive).
+func (r *Range) Begin() uint64 {
+	return r.begin
+}
+
+// End returns the last index covered by the range (exclusive).
+func (r *Range) End() uint64 {
+	return r.end
+}
+
+// Hashes returns sub-tree hashes corresponding to the minimal set of perfect
+// sub-trees covering the [begin, end) range, ordered left to right.
+func (r *Range) Hashes() [][]byte {
+	return r.hashes
+}
+
+// Append extends the compact range by appending the passed in hash to it. It
+// uses the tree hasher to calculate hashes of newly created nodes, and reports
+// them through the visitor function (if non-nil).
+func (r *Range) Append(hash []byte, visitor VisitFn) error {
+	return r.appendImpl(r.end+1, hash, nil, visitor)
+}
+
+// AppendRange extends the compact range by merging in the other compact range
+// from the right. It uses the tree hasher to calculate hashes of newly created
+// nodes, and reports them through the visitor function (if non-nil).
+func (r *Range) AppendRange(other *Range, visitor VisitFn) error {
+	if other.f != r.f {
```

Comment:
I guess pointer equality is probably mostly fine, though I wonder if there's a stronger notion which could be encoded in an `Equals`?

Reply:
SGTM